### DISCUSSION TITLE
Unknown attribute error

### DISCUSSION THREAD

**Discussion by Beyley** [gfx-rs/wgpu#3152] [Q&A] [answered]
I am attempting to port the https://github.com/gfx-rs/wgpu-native/tree/master/examples/compute project to C# as an example for the WIP Silk.NET bindings, but upon attempting to compile the shader i get an error claiming '@compute' is an invalid attribute, even though every example of WGSL compute ive seen shows that attribute being used, is this possibly some problem with the bindings or some breaking change in wgpu-native and the example hasnt been updated to address it?

Code:
https://github.com/dotnet/Silk.NET/blob/7984f8aabf08c765c56f4f7f0e5eaf35e5ce2c49/src/Lab/Experiments/WebGPUTest/Program.cs

Error: `Error: Internal Message: Parsing(ShaderError { source: "@group(0)\n@binding(0)\nvar<storage, read_write> v_indices: array<u32>; // this is used as both input and output for convenience\n\n// The Collatz Conjecture states that for any integer n:\n// If n is even, n = n/2\n// If n is odd, n = 3n+1\n// And repeat this process for each new n, you will always eventually reach 1.\n// Though the conjecture has not been proven, no counterexample has ever been found.\n// This function returns how many times this recurrence needs to be applied to reach 1.\nfn collatz_iterations(n_base: u32) -> u32{\n    var n: u32 = n_base;\n    var i: u32 = 0u;\n    loop {\n        if (n <= 1u) {\n            break;\n        }\n        if (n % 2u == 0u) {\n            n = n / 2u;\n        }\n        else {\n            // Overflow? (i.e. 3*n + 1 > 0xffffffffu?)\n            if (n >= 1431655765u) {   // 0x55555555u\n                return 4294967295u;   // 0xffffffffu\n            }\n\n            n = 3u * n + 1u;\n        }\n        i = i + 1u;\n    }\n    return i;\n}\n\n@compute\n@workgroup_size(1)\nfn main(@builtin(global_invocation_id) global_id: vec3<u32>) {\n    v_indices[global_id.x] = collatz_iterations(v_indices[global_id.x]);\n}", label: Some("Test Compute Shader"), inner: ParseError { message: "unknown attribute: 'compute'", labels: [(975..982, "unknown attribute")], notes: [] } })`

**Comment by ErichDonGubler** [comment, marked as answer]
TL;DR: I would bet that if you use a `master` build of `wgpu-native`, you'll be able to compile this shader as you expect.

---

## Questions

I gave up trying to repro this issue, because of these obstacles:

1. It's not clear to me what instructions should be followed for reproducing this locally with the given Silk.NET example. The translation looks more or less correct to me, but I wasn't able to determine how to run it.
2. It's not clear how you're consuming `wgpu-native`. It looks like bindings are being generated via NUKE (which is a build tool I'm not familiar with), but it's unclear where the headers used to generate `*.gen.cs` files and corresponding libs are coming from.

Without these, there's several leads to a fix that I see that can't be tested:

1. What version of `wgpu-native` are you consuming? Is there maybe a version mismatch? This seems likely, since `naga` (which handles WGSL input) 0.10 is the first version of W
2. Is there maybe an issue with bindings being generated?
3. Is this an environmental issue?

Bonus question: Why not file this discussion in `wgpu-native`, which is the layer you're actually depending on?

---

## Thoughts

*  The `@compute` attribute is specified by the [WGSL spec](https://gpuweb.github.io/gpuweb/wgsl/#attribute-compute):

    > Declares the function to be an [entry point](https://gpuweb.github.io/gpuweb/wgsl/#entry-point) for the [compute shader stage](https://gpuweb.github.io/gpuweb/wgsl/#compute-shader-stage) of a [compute pipeline](https://gpuweb.github.io/gpuweb/wgsl/#gpucomputepipeline).

    So, it's definitely _supposed_ to be supported at time of writing. I even tried copying over the version in your linked source to see if that caused any problems. I saw no difference in output (looks like the only changes are a few newlines removed, compared to the original).

* The `compute` example works locally on my Macbook Pro with M1 silicon:

    ```sh
    # Check out `wgpu-native`.
    # I tested on the current `master` branch: f3228dc9ad19dc4cbe43eff65ec532e6f94f7e9f.
    $ cd wgpu-native 

    $ cargo build
    <snip>
        Finished dev [unoptimized + debuginfo] target(s) in 0.07s

    $ cd examples/compute
    $ cmake . && make
    Times: [0, 1, 7, 2]
    ```

    N.B. that the latest commit in `wgpu-native`'s `master` branch is consuming [`wgpu` 0.14](https://github.com/gfx-rs/wgpu/releases/tag/v0.14.0) and [`naga` 0.10](https://github.com/gfx-rs/naga/releases/tag/v0.10.0), which are very recent (ATOW). AFAICT, the consumption of these versions hasn't yet been released in `wgpu-native` yet.

**Reply by Beyley** [reply]
> TL;DR: I would bet that if you use a `master` build of `wgpu-native`, you'll be able to compile this shader as you expect.
> ## Questions
> 
> I gave up trying to repro this issue, because of these obstacles:
> 
>     1. It's not clear to me what instructions should be followed for reproducing this locally with the given Silk.NET example. The translation looks more or less correct to me, but I wasn't able to determine how to run it.
> 
>     2. It's not clear how you're consuming `wgpu-native`. It looks like bindings are being generated via NUKE (which is a build tool I'm not familiar with), but it's unclear where the headers used to generate `*.gen.cs` files and corresponding libs are coming from.
> 
> 
> Without these, there's several leads to a fix that I see that can't be tested:
> 
>     1. What version of `wgpu-native` are you consuming? Is there maybe a version mismatch? This seems likely, since `naga` (which handles WGSL input) 0.10 is the first version of W
> 
>     2. Is there maybe an issue with bindings being generated?
> 
>     3. Is this an environmental issue?
> 
> 
> Bonus question: Why not file this discussion in `wgpu-native`, which is the layer you're actually depending on?
> ## Thoughts
> 
>     * The `@compute` attribute is specified by the [WGSL spec](https://gpuweb.github.io/gpuweb/wgsl/#attribute-compute):
>       > Declares the function to be an [entry point](https://gpuweb.github.io/gpuweb/wgsl/#entry-point) for the [compute shader stage](https://gpuweb.github.io/gpuweb/wgsl/#compute-shader-stage) of a [compute pipeline](https://gpuweb.github.io/gpuweb/wgsl/#gpucomputepipeline).
>       
>       
>       So, it's definitely _supposed_ to be supported at time of writing. I even tried copying over the version in your linked source to see if that caused any problems. I saw no difference in output (looks like the only changes are a few newlines removed, compared to the original).
> 
>     * The `compute` example works locally on my Macbook Pro with M1 silicon:
>       ```shell
>       # Check out `wgpu-native`.
>       # I tested on the current `master` branch: f3228dc9ad19dc4cbe43eff65ec532e6f94f7e9f.
>       $ cd wgpu-native 
>       
>       $ cargo build
>       <snip>
>           Finished dev [unoptimized + debuginfo] target(s) in 0.07s
>       
>       $ cd examples/compute
>       $ cmake . && make
>       Times: [0, 1, 7, 2]
>       ```
>       
>       
>           
>             
>           
>       
>             
>           
>       
>           
>         
>       N.B. that the latest commit in `wgpu-native`'s `master` branch is consuming [`wgpu` 0.14](https://github.com/gfx-rs/wgpu/releases/tag/v0.14.0) and [`naga` 0.10](https://github.com/gfx-rs/naga/releases/tag/v0.10.0), which are very recent (ATOW). AFAICT, the consumption of these versions hasn't yet been released in `wgpu-native` yet.

My bad for not being clear about running the app locally, my WebGPU stuff is very WIP currently, and yes since you mentioned it, its very possible its a mismatch of library versions, i downloaded the latest version from github releases, which may be out of date compared to the master branch of [webgpu-headers](https://github.com/webgpu-native/webgpu-headers/) (which are the headers im parsing to generate the bindings)
And to answer the bonus question: i hadn't filed this under wgpu-native or webgpu-headers as after reading the code, it seemed more possible to be a problem with the underlying wgpu code, but i wasn't totally sure, so i just filed it here, i shall try building wgpu-native from source and come back when i have results

**Reply by Beyley** [reply]
Doh yeah you were right! i built wgpu-native from source and it seems to work now! thank you for helping me with my stupidity

**Reply by ErichDonGubler** [reply]
> thank you for helping me with my stupidity

Trying the first thing that seems like it should work isn't stupid! WGSL is still technically a moving target right now, so while it's a bummer that you had an issue, it isn't necessarily surprising. I'm happy that you were able to get a fast resolution. :)
